src/conn/conn_tiered.c

```diff
@@ -121,55 +121,54 @@ __flush_tier_once(WT_SESSION_IMPL *session, uint32_t flags)
-    len = strlen("file:") + strlen(filename) + 1;
-    WT_ERR(__wt_calloc_def(session, len, &newfile));
-    WT_ERR(__wt_snprintf(newfile, len, "file:%s", filename));
-
-    /*
-     * If the file:URI of the tiered object does not exist, there is nothing to do.
-     */
-    ret = __wt_metadata_search(session, newfile, &config);
-    if (ret == WT_NOTFOUND) {
-        ret = 0;
-        goto err;
-    }
-    WT_ERR(ret);
+    const char *object;
 
-    /*
-     * We have a local version of this tiered data. Check its metadata for when it expires and
-     * remove if necessary.
-     */
-    cfg[0] = config;
-    cfg[1] = NULL;
-    WT_ERR(__wt_config_gets(session, cfg, "local_retention", &cval));
-    __wt_seconds(session, &now);
-    if (force || (uint64_t)cval.val + S2C(session)->bstorage->retain_secs >= now)
+    entry = NULL;
+    for (;;) {
+        __wt_seconds(session, &now);
+        __wt_tiered_get_drop_local(session, now, &entry);
+        if (entry == NULL)
+            break;
+        WT_ERR(__wt_tiered_name(
+          session, &entry->tiered->iface, entry->id, WT_TIERED_NAME_OBJECT, &object));
+        __wt_verbose(session, WT_VERB_TIERED, "REMOVE_LOCAL: %s at %" PRIu64, object, now);
+        WT_PREFIX_SKIP_REQUIRED(session, object, "object:");
         /*
-         * We want to remove the entry and the file. Probably do a schema_drop on the file:uri.
+         * If the handle is still open, it could still be in use for reading. In that case put the
+         * work unit back on the work queue and keep trying.
          */
-        __wt_verbose(session, WT_VERB_TIERED, "Would remove %s. Local retention expired", newfile);
-
+        if (__wt_handle_is_open(session, object)) {
+            __wt_verbose(session, WT_VERB_TIERED, "REMOVE_LOCAL: %s in USE, queue again", object);
+            WT_STAT_CONN_INCR(session, local_objects_inuse);
+            /*
+             * FIXME-WT-7470: If the object we want to remove is in use this is the place to call
+             * object sweep to clean up block->ofh file handles. Another alternative would be to try
+             * to sweep and then try the remove call below rather than pushing it back on the work
+             * queue. NOTE: Remove 'ofh' from s_string.ok when removing this comment.
+             */
+            __wt_tiered_push_work(session, entry);
+        } else {
+            __wt_verbose(session, WT_VERB_TIERED, "REMOVE_LOCAL: actually remove %s", object);
+            WT_STAT_CONN_INCR(session, local_objects_removed);
+            WT_ERR(__wt_fs_remove(session, object, false));
+            /*
+             * We are responsible for freeing the work unit when we're done with it.
+             */
+            __wt_tiered_work_free(session, entry);
+            entry = NULL;
+        }
+    }
 err:
-    __wt_free(session, config);
-    __wt_free(session, newfile);
+    if (entry != NULL)
+        __wt_tiered_work_free(session, entry);
```

Comment:
The only way to break out of the loop is with entry NULL because there is no other work or we get an error with a valid entry. However typing that makes me think we need to check for a panic connection too.

Reply:
Fixed